app/models/user.rb

```diff
@@ -22,54 +22,79 @@ class User < ActiveRecord::Base
 
   validates :email, uniqueness: true, allow_nil: true,
-    format: { with: /\A[^@\s]+@[^@\s]+\Z/ }
+                    format: {with: /\A[^@\s]+@[^@\s]+\Z/}
 
-  validates :title, :inclusion => { :in => TITLE_OPTIONS }, :allow_nil => true
+  validates :title, :inclusion => {:in => TITLE_OPTIONS}, :allow_nil => true
 
   after_create :generate_auth_token
 
-  scope :reviewers,   -> { where(roles: { name: 'Reviewer' }).joins(:roles) }
-  scope :supervisors, -> { where(roles: { name: 'Supervisor' }).joins(:roles) }
-  scope :order_fulfilment, -> { where(roles: { name: 'Order fulfilment' }).joins(:roles) }
-  scope :system,      -> { where(roles: { name: 'System' }).joins(:roles) }
-  scope :staff,       -> { where(roles: { name: ['Supervisor', 'Reviewer'] }).joins(:roles) }
+  scope :reviewers, -> { where(roles: {name: "Reviewer"}).joins(:roles) }
+  scope :supervisors, -> { where(roles: {name: "Supervisor"}).joins(:roles) }
+  scope :order_fulfilment, -> { where(roles: {name: "Order fulfilment"}).joins(:roles) }
+  scope :system, -> { where(roles: {name: "System"}).joins(:roles) }
+  scope :staff, -> { where(roles: {name: ["Supervisor", "Reviewer"]}).joins(:roles) }
   scope :except_stockit_user, -> { where.not(first_name: "Stockit", last_name: "User") }
 
   # used when reviewer is logged into donor app
   attr :treat_user_as_donor
 
   #added to allow sign_up without mobile number from stock app.
-  attr_accessor :request_from_stock
+  attr_accessor :request_from_stock, :request_from_browse
 
   # If user exists, ignore data and just send_verification_pin
   # Otherwise, create new user and send pin
   def self.creation_with_auth(user_params, app_name)
-    mobile = user_params['mobile']
-    user = find_by_mobile(mobile) if mobile.present?
+    mobile = user_params["mobile"].presence
+    email = user_params["email"].presence
+    user = find_user_by_mobile_or_email(mobile, email)
     user ||= new(user_params)
+    user.request_from_browse = (app_name == BROWSE_APP)
+    user.save if user.changed?
+    user.send_verification_pin(app_name, mobile, email) if user.valid?
+    user
+  end
+
+  def self.find_user_by_mobile_or_email(mobile, email)
+    return find_by_mobile(mobile) if mobile
+    find_by_email(email) if email
+  end
+
+  def send_sms(app_name)
     begin
-      user.save if user.changed?
-      user.send_verification_pin(app_name) if user.valid?
+      TwilioService.new(self).sms_verification_pin(app_name)
     rescue Twilio::REST::RequestError => e
-      msg = e.message.try(:split, '.').try(:first)
-      user.errors.add(:base, msg)
+      msg = e.message.try(:split, ".").try(:first)
+      self.errors.add(:base, msg)
+    end
+  end
+
+  def send_pin_email
+    begin
+      SendgridService.new(self).send_pin_email
+    rescue => e
+      Rollbar.error(e, error_class: "Sendgrid Error", error_message: "Sendgrid pin email")
     end
-    user
+  end
+
+  def send_verification_pin(app_name, mobile, email)
```

Comment:
@namrataukirde question :
 
If I sign-up with my email, and later add my mobile information. Will all the pin codes switch to mobile ?
Is that an expected behaviour ? I don't really know what's the best experience in that case, just thinking out loud. Maybe the social workers will still prefer the email 🤔

Reply:
Whaver user uses at the time of login then mode will be considered for sending pin. Like if user has both mobile and email_id but uses mobile number for login then sms will be send. If user uses email_id for login then we will send email for pin.

Reply:
oh I see, you pass in the payload data to the function. got it